syncstorage-spanner/src/error.rs

```diff
@@ -0,0 +1,145 @@
+use std::fmt;
+
+use backtrace::Backtrace;
+use http::StatusCode;
+use syncserver_common::{from_error, impl_fmt_display, ReportableError};
+use syncstorage_db_common::error::{DbErrorIntrospect, SyncstorageDbError};
+use thiserror::Error;
+
+/// An error type that represents any Spanner-related errors that may occur while processing a
+/// syncstorage request. These errors may be application-specific or lower-level errors that arise
+/// from the database backend.
+#[derive(Debug)]
+pub struct DbError {
+    kind: DbErrorKind,
+    pub status: StatusCode,
+    pub backtrace: Backtrace,
+}
+
+impl DbError {
+    pub fn batch_not_found() -> Self {
+        DbErrorKind::Common(SyncstorageDbError::batch_not_found()).into()
+    }
+
+    pub fn bso_not_found() -> Self {
+        DbErrorKind::Common(SyncstorageDbError::bso_not_found()).into()
+    }
+
+    pub fn collection_not_found() -> Self {
+        DbErrorKind::Common(SyncstorageDbError::collection_not_found()).into()
+    }
+
+    pub fn conflict() -> Self {
+        DbErrorKind::Common(SyncstorageDbError::conflict()).into()
+    }
+
+    pub fn expired() -> Self {
+        DbErrorKind::Expired.into()
+    }
+
+    pub fn integrity(msg: String) -> Self {
+        DbErrorKind::Integrity(msg).into()
+    }
+
+    pub fn internal(msg: String) -> Self {
+        DbErrorKind::Common(SyncstorageDbError::internal(msg)).into()
+    }
+
+    pub fn quota() -> Self {
+        DbErrorKind::Common(SyncstorageDbError::quota()).into()
+    }
+
+    pub fn too_large(msg: String) -> Self {
+        DbErrorKind::TooLarge(msg).into()
+    }
+}
+
+#[derive(Debug, Error)]
+enum DbErrorKind {
+    #[error("{}", _0)]
+    Common(SyncstorageDbError),
+
+    #[error("Connection expired")]
+    Expired,
+
+    #[error("A database error occurred: {}", _0)]
+    Grpc(#[from] grpcio::Error),
+
+    #[error("Database integrity error: {}", _0)]
+    Integrity(String),
+
+    #[error("Spanner data load too large: {}", _0)]
+    TooLarge(String),
```

Comment:
These could have been pulled into a `SpannerDbError` type akin to the `MysqlDbError` type, but the reason I made the `MysqlDbError` type was because it's also used by Tokenserver